experimental/ast/decl_range.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+import (
+	"slices"
+
+	"github.com/bufbuild/protocompile/internal/arena"
+)
+
+// DeclRange represents an extension or reserved range declaration. They are almost identical
+// syntactically so they use the same AST node.
+//
+// In the Protocompile AST, ranges can contain arbitrary expressions. Thus, DeclRange
+// implements [Comma[Expr]].
+type DeclRange struct {
+	withContext
+
+	ptr arena.Untyped
+	raw *rawDeclRange
+}
+
+type rawDeclRange struct {
+	keyword rawToken
+	args    []struct {
+		expr  rawExpr
+		comma rawToken
+	}
```

Comment:
The last part isn't really doable, if only because we have many `[]commaElem[T]` that are implementing a `Commas[U]`, for e.g. T = rawExpr and U = Expr.